### PR TITLE
Add WPML support to the sitemaps module

### PR DESCRIPTION
**Pull request by vukvukovich**
Fixes # .
#### Changes proposed in this Pull Request:

Sitemaps are now cached per language if WPML is active. 
Also, post permalinks are now generated properly according post language.
#### Testing instructions:
1. Install and activate **WP,** **WPML (whole bundle)** and **Jetpack**
2. Set up few languages, finish WPML configuration & enable Jetpack sitemaps module
3. Translate "Hello World" post to all languages
4. Scan Jetpack plugin for strings in **WPML > Theme and plugins localization**
5. Go to **WPML > String translation** and translate strings related to the sitemaps module. (you can find them by visiting _http://example.com/sitemap.xml_ or _http://example.com/news-sitemap.xml_)
6. Generate sitemaps for all languages by visiting _http://example.com/sitemap.xml_, _http://example.com/de/sitemap.xml_, etc...

**\- All strings should appear translated.** 
**\- All posts and its translations should be generated with proper permalinks.**

screen shot 2016-08-17 at 5 20 08 pm
screen shot 2016-08-17 at 5 20 21 pm
screen shot 2016-08-17 at 5 20 32 pm

---
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If [Gulp](http://gulpjs.com/) is installed on your testing environment, run `gulp js:hint` before to commit your changes. It will allow you to [detect errors in Javascript files](http://jshint.com/about/).
- [ ] Did you create a **new action or filter**? [Add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to help others understand how to use the action or the filter.
- [ ] Create **[unit tests](https://github.com/Automattic/jetpack/tree/master/tests)** if you can. If you're not familiar with Unit Testing, you can check [this tutorial](https://pippinsplugins.com/series/unit-tests-wordpress-plugins/).

### CODE REVIEW

File: modules/sitemaps/sitemaps.php [hunk 1]
```diff
@@ -202,6 +202,9 @@ function jetpack_load_xsl( $type = '' ) {
 
 	$transient_xsl = empty( $type ) ? 'jetpack_sitemap_xsl' : "jetpack_{$type}_sitemap_xsl";
 
+	// Add current language as suffix if WPML is active
+	$transient_xsl = $transient_xsl . wpml_transient_language();
```
Comment (kraftbj) [inline modules/sitemaps/sitemaps.php]: No function exists check here. Personally, I'd rather see modifications in https://github.com/Automattic/jetpack/blob/master/3rd-party/wpml.php using filters to add them.

**Comment by kraftbj** [comment]
Canceled the unit testing. Can you add this functionality via filters using /3rd-party/wpml.php instead? If there are additional filters needed in order to do it, we can look at adding them into sitemaps directly.

File: modules/sitemaps/sitemaps.php (hunk 1 above)
Comment (vukvukovich) [inline modules/sitemaps/sitemaps.php]: Function exists is not needed here, function is available at line 807.

Comment (vukvukovich) [inline modules/sitemaps/sitemaps.php]: Probably possible but you'll need a lot of hooks added.

Comment (vukvukovich) [inline modules/sitemaps/sitemaps.php]: Ok, is jetpack_wpml_transient_language() good or to remove wpml completely from equation? :)

Comment (kraftbj) [inline modules/sitemaps/sitemaps.php]: That's a fine function name, but still would need to put this in the compat file. It's going to get messy to have inline fixes for specific plugins. Just this one isn't terrible, but then Polylang will need something, then something will want something, etc.

Could have a filter for the transient name and likely use existing filters as hooks to call the language switch functions.

Comment (vukvukovich) [inline modules/sitemaps/sitemaps.php]: Totally agree, should I add filters or your team will?

Comment (kraftbj) [inline modules/sitemaps/sitemaps.php]: Please do—you can update this PR to do both add those filters and the functionality in the compat file.

If you could, please do add inline docs for the filters—that'd help the review process get knocked out faster! Thank you!

Comment (vukvukovich) [inline modules/sitemaps/sitemaps.php]: Sure, no problem. I'll update PR asap.

**Comment by vukvukovich** [comment]
That is it. I've tested it and it works properly.

There is other issue though, it is connected to this one, but we should deal with it in other PR. 

**3rd-party/wpml.php** is executing before WPML plugin. Which means, non of compatibility stuff is triggering. Did not test for other plugins, but I guess it is the same situation. 

I did fixed it locally just for sake of tests, but wanted to talk about it first, and see how do you want it to be handled.

Imho all compatibility files should kick in later.

Cheers!

**Comment by vukvukovich** [comment]
Do you need anything else from me here? Asking because of the label :)

**Comment by jeherve** [comment]
I updated the label :)

**Comment by vukvukovich** [comment]
Thanks @jeherve !

File: 3rd-party/wpml.php [hunk 2]
```diff
@@ -32,4 +32,51 @@ function grunion_contact_form_field_html_filter( $r, $field_label, $id ){
...
+/**
+ * Add current language as suffix to transient name.
+ *
+ * @param string $transient Transient name. Expected to not be SQL-escaped.
+ * @return string Translated transient name.
+ */
+add_filter( 'jetpack_sitemap_transient', 'jetpack_wpml_transient_language', 10, 1 );
+add_filter( 'jetpack_news_sitemap_transient', 'jetpack_wpml_transient_language', 10, 1 );
+add_filter( 'jetpack_sitemap_xsl_transient', 'jetpack_wpml_transient_language', 10, 1 ) ;
+
```
Comment (eliorivero) [inline 3rd-party/wpml.php]: Can you please move the inline doc that's written a few lines above to this location?

File: 3rd-party/wpml.php [hunk 3]
```diff
@@ -32,4 +32,51 @@ function grunion_contact_form_field_html_filter( $r, $field_label, $id ){
...
+/**
+ * Clear the multilingual sitemap cache when a sitemap action has changed.
+ *
+ * @param string $transient Transient name. Expected to not be SQL-escaped.
+ */
+add_action( 'delete_transient_jetpack_sitemap', 'jetpack_wpml_transient_delete', 10, 1 ) ;
+add_action( 'delete_transient_jetpack_news_sitemap', 'jetpack_wpml_transient_delete', 10, 1 );
+
+function jetpack_wpml_transient_delete( $transient ) {
```
Comment (eliorivero) [inline 3rd-party/wpml.php]: Also here, the inline doc written a few lines above must be moved to this location.

File: 3rd-party/wpml.php [hunk 4]
```diff
@@ -32,4 +32,51 @@ function grunion_contact_form_field_html_filter( $r, $field_label, $id ){
...
+
+/**
+ * Translate permalink according to post language.
+ *
+ * @param int $post_id Post ID.
+ * @return string|false The permalink URL
+ */
+add_filter( 'jetpack_sitemap_post_permalink', 'jetpack_wpml_translate_permalink', 10, 2 );
+
+function jetpack_wpml_translate_permalink( $permalink, $post_id ) {
```
Comment (eliorivero) [inline 3rd-party/wpml.php]: Here too, please, move the inline doc to this location.

File: modules/sitemaps/sitemaps.php [hunk 5]
```diff
@@ -202,6 +202,13 @@ function jetpack_load_xsl( $type = '' ) {
 
 	$transient_xsl = empty( $type ) ? 'jetpack_sitemap_xsl' : "jetpack_{$type}_sitemap_xsl";
 
+	/**
+	 * Filter transient name.
+	 *
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: The `$transient_xsl` parameter must be added in the inline doc.

File: modules/sitemaps/sitemaps.php [hunk 6]
```diff
@@ -250,7 +257,14 @@ function jetpack_print_news_sitemap_xsl() {
 function jetpack_print_sitemap() {
 	global $wpdb, $post;
 
-	$xml = get_transient( 'jetpack_sitemap' );
+	$xml = get_transient(
+		/**
+		 * Filter transient name.
+		 *
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: The parameter must be added in the inline doc give it a variable name.

File: modules/sitemaps/sitemaps.php [hunk 7]
```diff
@@ -443,7 +467,16 @@ function jetpack_print_sitemap() {
 	$xml = $tree->asXML();
 	unset( $tree );
 	if ( ! empty( $xml ) ) {
-		set_transient( 'jetpack_sitemap', $xml, DAY_IN_SECONDS );
+		set_transient(
+			/**
+			 * Filter transient name.
+			 *
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: Needs the parameter too.

File: modules/sitemaps/sitemaps.php [hunk 8]
```diff
@@ -460,7 +493,14 @@ function jetpack_print_sitemap() {
  */
 function jetpack_print_news_sitemap() {
 
-	$xml = get_transient( 'jetpack_news_sitemap' );
+	$xml = get_transient(
+		/**
+		 * Filter transient name.
+		 *
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: Needs the parameter.

File: modules/sitemaps/sitemaps.php [hunk 9]
```diff
@@ -601,7 +651,16 @@ function jetpack_print_news_sitemap() {
 	$xml = ob_get_contents();
 	ob_end_clean();
 	if ( ! empty( $xml ) ) {
-		set_transient( 'jetpack_news_sitemap', $xml, DAY_IN_SECONDS );
+		set_transient(
+			/**
+			 * Filter transient name.
+			 *
+			 * @module sitemaps
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: Needs the parameter.

File: modules/sitemaps/sitemaps.php [hunk 10]
```diff
@@ -706,6 +765,9 @@ function jetpack_sitemap_initialize() {
 	add_action( 'publish_page', 'jetpack_sitemap_handle_update', 12, 1 );
 	add_action( 'trash_post', 'jetpack_sitemap_handle_update', 12, 1 );
 	add_action( 'deleted_post', 'jetpack_sitemap_handle_update', 12, 1 );
+	// Clear sitemap cache if permalink structure is changed, used mostly because of multilingual plugins.
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: Whenever we add an inline doc, we add a line break before so it's easier to spot when scanning the file. Can you please add it?

**Comment by eliorivero** [comment]
Hi @vukvukovich, thanks for this. I've spend a fair amount of time reviewing this and here's what I found:

1) without this patch:
- the URLs are written correctly, although those corresponding to a certain language were duplicated. I'd expect to add those for a certain language only once.
captura de pantalla 2016-11-10 a las 18 24 57
- string translation works
- no transients by language since that's what this implements.

2) with this patch:
here's where things get tricky. Since you mentioned in a comment that `3rd-party/wpml.php` kicks in before WPML, I checked it out, and indeed, it kicks in when `ICL_SITEPRESS_VERSION` constant isn't defined so all the code wrapped in it isn't added. I'm going to add a PR to introduce a solution for this and will mention, it would be great if you could check it being a WPML dev.
So, I tweaked things so `3d-party/wpml.php` starts at the correct time and found that the URLs were still duplicated for those corresponding to a certain language, but this time, without the language slug.
captura de pantalla 2016-11-10 a las 18 24 31
The same thing happens for the sitemap in the original language:
captura de pantalla 2016-11-10 a las 18 29 54
so this should be addressed.

Transients per language are now written, so that's cool!
captura de pantalla 2016-11-10 a las 18 24 04

It would be good to generate the sitemap including only the entries written in the language requested each time. I notice that in your sitemap for `es` you still have the original entry and the one for for `sr`. That's the only one thing this PR is missing. Do you think you can introduce it?

Thanks!

**Comment by vukvukovich** [comment]
Sure, give me some time ti see and handle it that way. 

So, to sum up, we want only posts for requested language to be shown, right?

**Comment by eliorivero** [comment]
Yes, in that way it will render a proper view of the site's content in a certain language.

Also, I addressed the issue of the `3rd-party/wpml.php` file being loaded before WPML, thanks for catching that, and opened a PR here https://github.com/Automattic/jetpack/pull/5588
Would be great to have your feedback there.

Thanks again for all the contribution! 👍

File: modules/sitemaps/sitemaps.php (hunk 6 above)
Comment (vukvukovich) [inline modules/sitemaps/sitemaps.php]: Hi, 

Sorry did not understand, do you want me to add new var? 

Warm regards

**Comment by vukvukovich** [comment]
Done, you can check it now.

**Comment by vukvukovich** [comment]
@jeherve Any update on this proposed fix? Thanks!

**Comment by jeherve** [comment]
@vukvukovich We'll review your changes and see if we can include your changes into the next major Jetpack release, 4.6. 4.5 should be released this week, it's consequently too late to include new changes now.

**Comment by vukvukovich** [comment]
This was rather long ago, let me also check first if everything is okay since I am not sure xml is properly created.

**Comment by vukvukovich** [comment]
Hi @jeherve @eliorivero,

I've checked this MR again since long time passed.

Suggestion by @eliorivero which I've implemented in this MR is not the best one imho.

With this MR we have different sitemaps per language which can look ok to end user, but it is not according [Google standards](https://support.google.com/webmasters/answer/2620865?hl=en) and it does not serve its purpose.

So to follow the [Google standards](https://support.google.com/webmasters/answer/2620865?hl=en) bigger changes are needed.

Please tell me what do you think and how do you with to proceed.

Best,
Vuk

**Comment by eliorivero** [comment]
Hi Vuk, thanks for the continued work on this. The whole purpose of this is to explain the site structure to search engines, so we must follow their rules by all means. Feel free to make any changes that will make the sitemap understandable to search engines.
Also, feel free to counter me with arguments, please do so. If back in November you had posted this doc I would have gladly accepted it. Reviewing the PRs from contributors is one of many diverse tasks I have and I don't know about everything so whenever you explain something I'll be happy to accept it, I'm open to everything.
Thanks again for the hard work!

**Comment by vukvukovich** [comment]
Thanks @eliorivero on prompt answer.

You are definitely right. I had some similar issues long ago with some other plugins and I've totally slipped from my mind standards that I've read long ago.

So yes, lets deal with it like it should. 

For start, you can tell me what do you think we should present via XSLT?

Since we need all translations and to provide full set of URLs in xml, maybe we should present them all and make one universal sitemap.xml for all languages. (of course localized one, but with same URLs)

What do you think?

**Comment by jeherve** [comment]
We've made quite a few changes to the Sitemaps module in the past few months, so this PR isn't up to date anymore unfortunately.

**Comment by vukvukovich** [comment]
@jeherve How do we proceed here since there is a quite refactoring needed for this to be up to standards?

**Comment by zinigor** [comment]
@vukvukovich this PR seems to have slipped under our radar for a while. We're planning to make another change to the Sitemaps module in this cycle that will fix #7265, after which we will rebase and adapt your work here to the new code. Thanks a lot for doing this, we really appreciate it!

**Comment by vukvukovich** [comment]
@zinigor Thanks Igor, I know it is outdated, I don't even remember everything related to it. 😂 
I remember one important thing though, that if used multilingual it is not following [google standards](https://support.google.com/webmasters/answer/2620865?hl=en). 

Now, as WPML is already adding necessary data to page head, maybe we could disable or have note for users about that. That would be great if you decide, should we make module fully multilingual or make it as usable but not fully operational, since sitemaps are not used only for search engines but are also used by users to indicate site content.

What do you think?

**Comment by oskosk** [comment]
@vukvukovich this PR is too old in many ways not. Given that you mentioned "I don't even remember everything related to it" I'm closing this one. 

Feel free to reopen and rebase if you have some time. Or dropping an issue stating the problem this PR solved would be fine too. Thank you and I'm sorry this got stalled

**Pull request closed without merging**